page_analyzer/app.py

```diff
@@ -1,8 +1,115 @@
-from flask import Flask, render_template
+import os
+from typing import Any, Dict
+from urllib.parse import urlparse
+
+import psycopg2
+from dotenv import load_dotenv
+from flask import (
+    Flask,
+    flash,
+    get_flashed_messages,
+    redirect,
+    render_template,
+    request,
+    url_for,
+)
+from validators import url
+
+load_dotenv()
+
 
 app = Flask(__name__)
+app.secret_key = os.getenv("SECRET_KEY")
+DATABASE_URL = os.getenv("DATABASE_URL")
+
+
+def connect_db() -> Any:
+    conn = psycopg2.connect(DATABASE_URL)
+    return conn
+
+
+def get_url_context(url_id: int) -> Dict[str, Any]:
+    conn = connect_db()
+    cur = conn.cursor()
+    cur.execute("SELECT name FROM urls WHERE id='%s';", (url_id,))
+    page_url = cur.fetchone()[0]
+    cur.execute("SELECT created_at FROM urls WHERE id='%s';", (url_id,))
+    created_at = cur.fetchone()[0]
+    cur.close()
+    conn.close()
+    context = {"id": url_id, "name": page_url, "created_at": created_at}
+    return context
+
+
+def validate(url_name: str) -> Dict[str, Any]:
+    errors = {}
+    if not url(url_name) or len(url_name) > 255:
+        errors["name"] = "Некорректный URL"
+    return errors
+
+
+def normalize(url_name: str) -> str:
+    p = urlparse(url_name)
+    normalized = p.scheme + "://" + p.netloc
+    if p.port:
+        port = ":" + str(p.port)
+        normalized = normalized.replace(port, "")
+    return normalized
 
 
 @app.route("/")
-def hello_world() -> str:
-    return render_template("index.html")
+def index() -> str:
+    message = None
+    return render_template("index.html", message=message)
+
+
+@app.route("/url", methods=["POST"])
+def post_url() -> Any:
+    url_name = request.form["url"]
+    errors = validate(url_name)
+    if errors:
+        flash(errors["name"], "danger")
+        messages = get_flashed_messages(with_categories=True)
+        return render_template("index.html", message=messages[0])
+    url_name = normalize(url_name)
+    conn = connect_db()
+    cur = conn.cursor()
+    cur.execute("SELECT name FROM urls;")
+    names = cur.fetchall()
+    for name in names:
+        if name[0] == url_name:
+            cur.execute("SELECT id FROM urls WHERE name=%s;", (url_name,))
+            url_id = cur.fetchone()[0]
+            cur.close()
+            conn.close()
+            flash("Страница уже существует", "info")
+            return redirect(url_for("get_url", id=url_id))
+    cur.execute("INSERT INTO urls (name) VALUES (%s);", (url_name,))
+    conn.commit()
+    cur.execute("SELECT id FROM urls WHERE name=%s;", (url_name,))
+    url_id = cur.fetchone()[0]
+    cur.close()
+    conn.close()
+    flash("Страница успешно дoбавлена", "success")
+    return redirect(url_for("get_url", id=url_id))
+
+
+@app.route("/urls/<int:id>")
+def get_url(id: int) -> str:
+    url_data = get_url_context(id)
+    messages = get_flashed_messages(with_categories=True)
+    return render_template("url.html", url=url_data, message=messages[0])
```

Comment:
not sure if using flash is the good idea here, I'd rather pass the `url_data` object into view as a parameter